### PR TITLE
Global scheduler - per-task transfer-aware policy

### PR DESCRIPTION
**Pull request by atumanov**
* Basic per-task transfer-aware policy
  - falls back to simple round-robin policy when (a) no object dependencies, (b) dependencies are not cached, (c) dependencies are args by value
* Subscribe to all object, task, and db_client notifications (two subscribes and one psubscribe)
* Cache all objects and their object locations
* Cache information about plasma -- photon association
* Expand global scheduler test
  - integration testing (involving plasma store, plasma manager, photon, global scheduler, and Redis)
  - testing with synchronizing object and task notifications
  - testing without synchronizing object and task notifications (OoO notification arrival handling)

### CODE REVIEW

File: src/common/object_info.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#ifndef OBJECT_H
+#define OBJECT_H
+#include "common.h"
+#include "object_info.h"
+
+#include <stdio.h>
+
+void object_id_print(object_id obj_id) {
```
Comment (robertnishihara) [inline src/common/object_info.c]: These methods `object_id_print` and `object_id_tostring` should be in `common.c`

File: src/common/state/redis.c [hunk 2]
```diff
@@ -555,9 +555,18 @@ object_id parse_subscribe_to_notifications_payload(
   int num_managers = (length - sizeof(object_id) - 1 - sizeof(data_size_value) -
                       1 - strlen("MANAGERS")) /
                      (1 + sizeof(db_client_id));
-  CHECK(length ==
+
+  int64_t rval = sizeof(object_id) + 1 + sizeof(data_size_value) + 1 +
+      strlen("MANAGERS") + num_managers * (1 + sizeof(db_client_id));
+  char buf[1024];
```
Comment (robertnishihara) [inline src/common/state/redis.c]: `buf` and `bufptr` aren't used for anything, right?

File: src/global_scheduler/global_scheduler.c [hunk 3]
```diff
@@ -83,17 +92,100 @@ void process_new_db_client(db_client_id db_client_id,
                            const char *aux_address,
                            void *user_context) {
   global_scheduler_state *state = (global_scheduler_state *) user_context;
+  char objidstr[3*sizeof(unique_id)];
+  int objidlen = sizeof(objidstr);
+  LOG_INFO("db client table callback for db client=%s",
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.c]: `obj_id_len` and `obj_id_str`

File: src/common/object_info.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+  }
+}
+
+/**
+ * Creates a hex string(dot-delimited bytes) representation of the given obj_id.
+ * @param obj_id object id
+ * @return dot-delimited hex string representation of the object id.
+ */
+char *object_id_tostring(object_id obj_id, char *outstr, int outstrlen) {
```
Comment (pcmoritz) [inline src/common/object_info.c]: consider using sha1_to_hex here

Comment (robertnishihara) [inline src/common/object_info.c]: object_id_to_string

File: src/global_scheduler/global_scheduler_algorithm.c [hunk 5]
```diff
@@ -15,18 +16,167 @@ void destroy_global_scheduler_policy(
   free(policy_state);
 }
 
-void handle_task_waiting(global_scheduler_state *state,
-                         global_scheduler_policy_state *policy_state,
-                         task *task) {
+void handle_task_nodep(global_scheduler_state *state,
```
Comment (pcmoritz) [inline src/global_scheduler/global_scheduler_algorithm.c]: avoid abbreviations

File: src/common/object_info.c (hunk 4 above)
Comment (robertnishihara) [inline src/common/object_info.c]: out_str_len

File: src/global_scheduler/global_scheduler.c [hunk 6]
```diff
@@ -83,17 +92,100 @@ void process_new_db_client(db_client_id db_client_id,
...
 
+/**
+ * Process notification about the new object information.
+ *
+ * @param object_id : id of the object with new location
+ * @param data_size: object size
+ * @param manager_count: the count of new locations for this object
+ * @param manager_vector: the vector with new Plasma Manager locations
+ * @param user_context: user context passed to the object_table_subscribe()
+ *
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.c]: remove this newline

File: src/global_scheduler/global_scheduler.c [hunk 7]
```diff
@@ -83,17 +92,100 @@ void process_new_db_client(db_client_id db_client_id,
...
+ * @return None
+ */
+void object_table_subscribe_callback(
+    object_id object_id,
+    int64_t data_size,
+    int manager_count,
+    OWNER const char *manager_vector[],
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.c]: the OWNER macro should not be here, i think

File: src/global_scheduler/global_scheduler_algorithm.h [hunk 8]
```diff
@@ -13,6 +13,12 @@
  *
  */
 
+typedef enum {
+  SCHED_ROUND_ROBIN = 1,
+  SCHED_TRANSFER_AWARE = 2,
+  SCHED_POLICY_MAX
+} global_scheduler_policy_t;
```
Comment (pcmoritz) [inline src/global_scheduler/global_scheduler_algorithm.h]: let's get rid of the _t

File: src/global_scheduler/test/test.py [hunk 9]
```diff
@@ -49,20 +87,28 @@ def setUp(self):
...
     # Start the global scheduler.
     self.p1 = global_scheduler.start_global_scheduler(redis_address, use_valgrind=USE_VALGRIND)
+    #time.sleep(0.5)
     # Start the Plasma store.
     plasma_store_name, self.p2 = plasma.start_plasma_store()
     # Start the Plasma manager.
     plasma_manager_name, self.p3, plasma_manager_port = plasma.start_plasma_manager(plasma_store_name, redis_address)
-    plasma_address = "{}:{}".format(node_ip_address, plasma_manager_port)
+    self.plasma_address = "{}:{}".format(node_ip_address, plasma_manager_port)
+    self.plasmaclient = plasma.PlasmaClient(plasma_store_name, plasma_manager_name)
```
Comment (pcmoritz) [inline src/global_scheduler/test/test.py]: plasma_client

File: src/global_scheduler/test/test.py [hunk 10]
```diff
@@ -88,22 +134,65 @@ def tearDown(self):
     # after we kill the global scheduler.
     self.redis_process.kill()
 
-  def test_redis_contents(self):
-    # DB_CLIENT_PREFIX is an implementation detail of ray_redis_module.c, so
-    # this must be kept in sync with that file.
-    DB_CLIENT_PREFIX = "CL:"
+  def get_plasmamgr_id(self):
```
Comment (pcmoritz) [inline src/global_scheduler/test/test.py]: get_plasma_manager_id

File: src/global_scheduler/test/test.py [hunk 11]
```diff
@@ -88,22 +134,65 @@ def tearDown(self):
...
+    ''' iterates over all the client table keys, gets the db_client_id for the client
+        with client_type matching plasma_manager. Strips the client table prefix.
+        Returns None if plasma_manager client not found.
+        Note that it returns db_client id for the first encountered plasma manager.
+        TODO(atumanov): write a separate function to get all plasma manager client IDs
+    '''
+    db_client_id = None
+
+    cli_lst = self.redis_client.keys("{}*".format(DB_CLIENT_PREFIX))
```
Comment (pcmoritz) [inline src/global_scheduler/test/test.py]: client_list

File: src/global_scheduler/global_scheduler_algorithm.c [hunk 12]
```diff
@@ -126,17 +127,18 @@ db_client_id get_photon_id(global_scheduler_state *state,
...
-  /* Check to make sure this photon_db_client_id matches one of the schedulers*/
-  local_scheduler *lsptr = NULL;
-  for (lsptr = (local_scheduler *) utarray_front(state->local_schedulers);
-       lsptr != NULL; lsptr = (local_scheduler *) utarray_next(
-                          state->local_schedulers, lsptr)) {
-    if (memcmp(&lsptr->id, &photon_id, sizeof(photon_id)) == 0) {
+  /* Check to make sure this photon_db_client_id matches one of the
+   * schedulers. */
+  local_scheduler *local_scheduler_ptr;
+  for (int i = 0; i < utarray_len(state->local_schedulers); ++i) {
```
Comment (atumanov) [inline src/global_scheduler/global_scheduler_algorithm.c]: actually this for loop change won't work. Subsequent if (!lsptr) check assumes that lsptr reaches NULL value if it finds no matching local schedulers after traversing the full array.

File: src/global_scheduler/global_scheduler.h [hunk 13]
```diff
@@ -11,19 +11,24 @@
 typedef struct {
   /** The ID of the local scheduler in Redis. */
   db_client_id id;
+  /** The number of tasks sent from the local scheduler to the global
+   *  scheduler. */
```
Comment (atumanov) [inline src/global_scheduler/global_scheduler.h]: Number of tasks sent from the global to the local

File: src/global_scheduler/global_scheduler.c [hunk 14]
```diff
@@ -53,13 +53,21 @@ void free_global_scheduler(global_scheduler_state *state) {
...
     HASH_DELETE(hh, state->plasma_photon_map, entry);
     /* Now deallocate hash table entry. */
     free(entry->aux_address);
     free(entry);
   }
+  /* Free the scheduler object info table. */
+  scheduler_object_info *object_entry, *tmp_entry;
+  HASH_ITER(hh, state->scheduler_object_info_table, object_entry, tmp_entry) {
+    HASH_DELETE(hh, state->scheduler_object_info_table, object_entry);
+    utarray_free(object_entry->object_locations);
```
Comment (atumanov) [inline src/global_scheduler/global_scheduler.c]: I think we need to iterate over object_locations and individually free each location. IIRC, utarray_push_back strdups the strings

Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.c]: Could be mistaken, but from this example code, it looks like we don't have to. http://www.cs.bu.edu/~jappavoo/Resources/psml/apps/hashy/uthash/doc/utarray.html#_strings

**Pull request merged**